elide-core/src/main/java/com/yahoo/elide/jsonapi/extensions/JsonApiAtomicOperations.java

```diff
@@ -0,0 +1,563 @@
+                }
+                JsonNode data = operation.getData();
+                String href = operation.getHref();
+                Ref ref = operation.getRef();
+                String fullPath = href;
+                if (fullPath == null) {
+                    if (ref == null) {
+                        ref = inferRef(requestScope.getMapper(), operation);
+                    }
+                    fullPath = getFullPath(ref, operation);
+
+                    // If data is not provided, but it is to remove a resource the data is optional
+                    // and should be generated from the ref
+                    if ((data == null || JsonNodeType.NULL.equals(data.getNodeType()))
+                            && OperationCode.REMOVE.equals(operation.getOperationCode())) {
+                        data = requestScope.getMapper().getObjectMapper().createObjectNode().put("type", ref.getType())
+                                .put("id", ref.getId() != null ? ref.getId() : ref.getLid());
+                    }
+                }
+                if (fullPath == null) {
+                    throw new InvalidEntityBodyException(
+                            "Atomic Operations extension requires either href or ref to be specified.");
+                }
+
+                switch (operation.getOperationCode()) {
+                    case ADD:
+                        result = handleAddOp(fullPath, data, requestScope, action);
+                        break;
+                    case UPDATE:
+                        result = handleUpdateOp(fullPath, data, requestScope, action);
+                        break;
+                    case REMOVE:
+                        result = handleRemoveOp(fullPath, data, requestScope);
+                        break;
+                    default:
+                        throw new InvalidEntityBodyException(
+                            "Invalid Atomic Operations extension operation code:"
+                                    + operation.getOperationCode());
+                }
+                return result;
+            } catch (HttpStatusException e) {
+                action.cause = e;
+                throw e;
+            }
+        }).toList();
+    }
+
+    /**
+     * Infer ref using the data for operations on resources. The ref cannot be
+     * inferred for operations on relationships.
+     *
+     * @param mapper the json api mapper
+     * @param operation the operation
+     * @return the ref
+     */
+    private Ref inferRef(JsonApiMapper mapper, Operation operation) {
+        // Attempt to infer the ref from the data
+        if (operation.getData() != null && !operation.getData().isArray()) {
+            try {
+                Resource resource = mapper.forAtomicOperations()
+                        .readResource(operation.getData());
+                if (resource.getType() != null) {
+                    if (resource.getAttributes() == null || resource.getAttributes().isEmpty()) {
```

Comment:
Why do we care about attributes being null/empty here?

Reply:
This is to tell whether it was an relationship operation. If it is a relationship operation and the attributes would be null or empty.

Reply:
The spec says if the operations is targeting a resource, then ref/href MUST be used.  The same goes for deleting a resource (non-relationship).  Should this be an error instead (instead of inferring the ref)?

Reply:
After looking at the spec again it looks like you are correct that for delete resource scenario the ref/href is mandatory. I have updated the code to only infer the ref for the add and update resource scenario where the ref/href is optional. The code thus will not attempt to infer for the delete resource and any relationship operations scenario.